### PR TITLE
feat: MCP tools, hooks, export, search threshold + fixes

### LINKED ISSUE #475: BUG: Float mtime comparison breaks file deduplication — every file re-mined on each run
## Summary

`file_already_mined()` in `palace.py:68` uses strict float equality to compare stored vs current mtime:

```python
return float(stored_mtime) == current_mtime
```

`os.path.getmtime()` returns a float. ChromaDB stores metadata via JSON serialization, which introduces floating-point precision loss (e.g., `1712345678.123456` may round-trip as `1712345678.1234560012817383`). The strict `==` comparison frequently fails even for unchanged files, causing **every file to be re-mined on every run**.

This defeats the entire dedup/skip mechanism and silently bloats the palace with duplicate drawers.

## Reproduction

1. `mempalace init <dir> && mempalace mine <dir>` — initial mine
2. `mempalace mine <dir>` — re-mine without changing any files
3. Observe that all files are processed again (not skipped)

## Suggested Fix

Use epsilon comparison:

```python
return abs(float(stored_mtime) - current_mtime) < 0.01
```

Or truncate to integer seconds:

```python
return int(float(stored_mtime)) == int(current_mtime)
```

## Environment

- mempalace 3.1.0
- ChromaDB 0.6.3
- Linux x86_64, Python 3.12

### LINKED ISSUE #477: BUG: MCP tool_search has no upper bound on limit parameter — potential memory exhaustion
## Summary

`tool_search` in `mcp_server.py` (line ~249) accepts `limit: int = 5` with no validation. The MCP schema declares it as `"type": "integer"` with no `minimum`/`maximum` constraints.

A caller (or a prompt-injected AI) can pass `limit=1000000`, which gets forwarded directly to `col.query(n_results=1000000)`. ChromaDB will attempt to allocate embeddings for all results, exhausting memory or hanging.

## Suggested Fix

Clamp the limit:

```python
limit = max(1, min(limit, 100))
```

And add constraints to the MCP schema:

```json
{"type": "integer", "default": 5, "minimum": 1, "maximum": 100}
```

## Related

Similar to the security hardening discussed in #401.

### PR DESCRIPTION
**Pull request by jphein**
## Summary
- **Mtime dedup fix** — epsilon comparison (`abs() < 0.01`) instead of float `==` (fixes #475)
- **MCP server fixes** — search limit capped [1,100], status/taxonomy tools paginated past 10K, duplicate cache declarations removed (fixes #477, #478, #479)
- **Similarity threshold** — `min_similarity` parameter in search, default 1.5 L2 distance in MCP
- **Hooks rewrite** — stop/precompact hooks now instruct AI to use mempalace MCP tools (`mempalace_diary_write`, `mempalace_add_drawer`), auto-ingest Claude Code JSONL transcripts
- **3 new MCP tools** — `get_drawer`, `list_drawers` (paginated), `update_drawer` (with WAL audit)
- **WAL chmod fix** — only chmod on file creation, not every write
- **Metadata cache** — 5s TTL for status/wings/taxonomy to avoid redundant pagination
- **Markdown export** — `mempalace export -o <dir>` renders palace as browsable markdown files
- **CLAUDE.md** — project documentation for Claude Code context

Supersedes #483 and #484 (includes those fixes plus new features).

## Test plan
- [x] 562 tests pass
- [ ] Verify `mempalace export -o /tmp/test` produces correct markdown tree
- [ ] Verify new MCP tools (`get_drawer`, `list_drawers`, `update_drawer`) via MCP protocol
- [ ] Verify stop hook triggers mempalace save instructions

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by web3guru888 — COMMENTED** [review]
Solid PR, jphein — this addresses several gaps I've hit in production use.

**Similarity threshold** — This is the feature I've wanted most. We implemented `min_similarity` filtering internally (threshold 0.72 for our use case) and it dramatically cuts noise on broad queries. One critical thing to document: is `min_similarity` filtering against L2 distance (higher = less similar) or cosine similarity (higher = more similar)? ChromaDB's default embedding function returns L2 by default, so a threshold of 1.5 is *excluding* scores above 1.5, not including. That's counterintuitive for users coming from cosine-similarity workflows and should be called out clearly in CLAUDE.md and/or the docstring. Easy fix but prevents a lot of confusion.

**Hooks rewrite** — Going through MCP tools (`mempalace_diary_write`, `mempalace_add_drawer`) instead of direct Python calls is the right architectural choice. Keeps hooks decoupled from internals and benefits automatically from validation, pagination, etc. The auto-ingest of Claude Code JSONL transcripts on compaction is clever — turns a normally lossy event into a structured memory commit.

**New MCP tools** — `list_drawers` with wing/room pagination fills a real gap. Right now you're flying blind unless you already know what to search for. `get_drawer` by ID is handy for follow-up fetches after a search result. Both straightforward, good additions.

**`export_palace` at scale** — Does this stream or paginate? If a palace has 10K–50K drawers (which is realistic for long-running integrations), a single export call risks timeout at the MCP layer. I'd suggest either: (a) returning a paginated export with `page`/`page_size` params, or (b) writing to a temp file on disk and returning the path. The exporter module design looks right for (b) if `export_palace` just wraps a file write.

**Conflict with #484** — Both this PR and #484 touch `mcp_server.py`. Worth checking if there's overlap before merge — specifically around the search tool signature changes.

Mtime epsilon fix is a clean correctness improvement. No concerns there.

Overall: 👍 on direction. Threshold docs + export scaling are the two things worth addressing before merge.

File: mempalace/cli.py [hunk 1]
```diff
@@ -95,6 +95,7 @@ def cmd_mine(args):
             dry_run=args.dry_run,
             respect_gitignore=not args.no_gitignore,
             include_ignored=include_ignored,
+            workers=args.workers,
         )
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/cli.py:99]: This is already working — `mine()` has the `workers` parameter (line 661). The PR diff may not show it since it was added in the concurrent mining commit.

File: mempalace/hooks_cli.py [hunk 2]
```diff
@@ -103,6 +108,37 @@ def _maybe_auto_ingest():
...
+    path = Path(transcript_path).expanduser()
+    if not path.is_file() or path.stat().st_size < 100:
+        return
+
+    from .config import MempalaceConfig
+
+    try:
+        palace_path = MempalaceConfig().palace_path
+    except Exception:
+        return
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/hooks_cli.py:244]: Fixed — changed to `MempalaceConfig()` (validate config loads) instead of assigning to unused `palace_path`.

File: tests/test_exporter.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+import os
+import shutil
+import tempfile
+from pathlib import Path
+
+import chromadb
+import yaml
```
[the remark this answers is not in the document]
Comment (jphein) [reply tests/test_exporter.py:6]: Fixed — removed unused `chromadb` import from test_exporter.py.

File: mempalace/exporter.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    for wing in sorted(grouped):
+        wing_dir = os.path.join(output_dir, wing)
+        os.makedirs(wing_dir, exist_ok=True)
+        wing_drawer_count = 0
+
+        rooms = grouped[wing]
+        for room in sorted(rooms):
+            drawers = rooms[room]
+            room_path = os.path.join(wing_dir, f"{room}.md")
+
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/exporter.py]: Fixed — added `_safe_path_component()` that strips `/\:*?"<>|` and leading dots from wing/room metadata before using as directory/file names. Applied to both wing dirs and room filenames in the streaming exporter.

File: mempalace/searcher.py [hunk 5]
```diff
@@ -142,11 +147,19 @@ def search_memories(
...
+                "distance": round(dist, 4),
             }
         )
 
+    # Filter out results exceeding the distance threshold.
+    # ChromaDB default L2: lower distance = more similar.
+    # min_similarity=0.0 (default) disables filtering for backwards compat.
+    if min_similarity > 0.0:
+        hits = [h for h in hits if h["distance"] <= min_similarity]
+
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/searcher.py]: Fixed — distance filtering now happens on the raw float before rounding, then the display values are rounded for output. Avoids borderline precision loss.

File: mempalace/mcp_server.py [hunk 6]
```diff
@@ -410,6 +457,136 @@ def tool_delete_drawer(drawer_id: str):
...
+            "metadata": meta,
+        }
+    except Exception as e:
+        return {"error": str(e)}
+
+
+def tool_list_drawers(wing: str = None, room: str = None, limit: int = 20, offset: int = 0):
+    """List drawers with pagination. Optional wing/room filter."""
+    limit = max(1, min(limit, 100))
+    col = _get_collection()
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/mcp_server.py:490]: Fixed — added `offset = max(0, offset)` at the top of `tool_list_drawers()`.

File: mempalace/mcp_server.py [hunk 7]
```diff
@@ -410,6 +457,136 @@ def tool_delete_drawer(drawer_id: str):
...
+                "content_changed": content is not None,
+                "content_preview": new_doc[:200] if content is not None else None,
+            },
+        )
+
+        update_kwargs = {"ids": [drawer_id]}
+        if content is not None:
+            update_kwargs["documents"] = [new_doc]
+        update_kwargs["metadatas"] = [new_meta]
+        col.update(**update_kwargs)
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/mcp_server.py:587]: Fixed — `tool_update_drawer()` now returns a no-op response when all optional params are omitted (no DB write). Also invalidates `_metadata_cache` after successful updates so status/taxonomy tools don't serve stale data.

File: mempalace/mcp_server.py [hunk 8]
```diff
@@ -734,14 +911,18 @@ def tool_diary_read(agent_name: str, last_n: int = 10):
...
     },
     "mempalace_search": {
-        "description": "Semantic search. Returns verbatim drawer content with similarity scores.",
+        "description": "Semantic search. Returns verbatim drawer content with similarity scores. Results with distance > min_similarity are filtered out (L2 distance: lower = more similar).",
         "input_schema": {
             "type": "object",
             "properties": {
                 "query": {"type": "string", "description": "What to search for"},
                 "limit": {"type": "integer", "description": "Max results (default 5)"},
                 "wing": {"type": "string", "description": "Filter by wing (optional)"},
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/mcp_server.py:998]: Fixed — added `minimum: 1, maximum: 100` to the search limit schema and list_drawers limit/offset schemas.

File: mempalace/mcp_server.py [hunk 9]
```diff
@@ -410,6 +457,136 @@ def tool_delete_drawer(drawer_id: str):
...
+            "offset": offset,
+            "limit": limit,
+        }
+    except Exception as e:
+        return {"error": str(e)}
+
+
+def tool_update_drawer(drawer_id: str, content: str = None, wing: str = None, room: str = None):
+    """Update an existing drawer's content and/or metadata."""
+    col = _get_collection()
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/mcp_server.py:539]: Noted — will add MCP tool tests for get/list/update_drawer in a follow-up commit. The existing test_mcp_server.py covers the other tools; these three need coverage for basic CRUD and pagination.

File: CLAUDE.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+- `~/.mempalace/config.json` — topic wings and hall keywords, customized for JP's domains (infrastructure, development, tools, creative, projects, system).
+- `~/.mempalace/palace/` — ChromaDB vector store. The actual data.
+- `~/.mempalace/hook_state/` — stop hook session tracking.
+
+## Development
+
+```bash
+source venv/bin/activate
+python -m pytest tests/ -x -q           # run tests (534 expected)
+mempalace status                         # check palace state
```
[the remark this answers is not in the document]
Comment (jphein) [reply CLAUDE.md:24]: Fixed — updated CLAUDE.md from 534 to 562 expected tests.

File: mempalace/searcher.py [hunk 11]
```diff
@@ -91,7 +91,12 @@ def search(query: str, palace_path: str, wing: str = None, room: str = None, n_r
...
 
 def search_memories(
-    query: str, palace_path: str, wing: str = None, room: str = None, n_results: int = 5
+    query: str,
+    palace_path: str,
+    wing: str = None,
+    room: str = None,
+    n_results: int = 5,
+    min_similarity: float = 0.0,
 ) -> dict:
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/searcher.py:104]: Agreed the naming is confusing. For now I've added comprehensive L2 distance documentation in the function docstring and MCP schema description. A full rename to `max_distance` would be a breaking change — happy to do it in a separate PR with a deprecation alias if you'd prefer.

File: mempalace/mcp_server.py [hunk 12]
```diff
@@ -794,6 +975,59 @@ def tool_diary_read(agent_name: str, last_n: int = 10):
...
+                "wing": {"type": "string", "description": "Filter by wing (optional)"},
+                "room": {"type": "string", "description": "Filter by room (optional)"},
+                "limit": {
+                    "type": "integer",
+                    "description": "Max results per page (default 20, max 100)",
+                },
+                "offset": {
+                    "type": "integer",
+                    "description": "Offset for pagination (default 0)",
+                },
```
[the remark this answers is not in the document]
Comment (jphein) [reply mempalace/mcp_server.py:1084]: Fixed — added `minimum`/`maximum` bounds to list_drawers schema: limit min 1 max 100, offset min 0.

**Review by web3guru888 — COMMENTED** [review]
Follow-up after jphein's fixes:

The Copilot-flagged issues are all resolved — especially the `_safe_path_component()` path traversal fix in the exporter and the no-op guard in `tool_update_drawer()`. Both of those were real correctness issues and the fixes are clean.

**On `min_similarity` naming**: The docstring approach is fine for now. The L2 distance direction clarification ("lower = more similar") is the most important part, and having it in both the docstring and MCP schema description covers both the code and the LLM context path. A follow-up PR with a `max_distance` alias + deprecation of `min_similarity` would be a good ergonomic improvement — our integration still trips over this occasionally when onboarding new contributors.

**On export scaling**: Glad the exporter writes to disk and returns a path rather than the full content inline. That's the right call for large palaces. One edge case worth a test: an export run while concurrent mining is happening — HNSW doesn't lock between reads and writes, so a large export could see a partially-committed collection if mining hits a batch boundary mid-export. A read-lock or a "stop mining first" doc note would cover this.

**On the CLAUDE.md test count drift**: Good call updating from 534 → 562. Worth making this dynamic (e.g., `pytest --co -q | tail -1`) in CI so it doesn't need manual updates going forward.

This is mergeable IMO. The three new MCP tool tests are technically a gap but the pattern is identical to existing tool tests and jphein has noted it as a follow-up — that's a reasonable trade for getting the other fixes in.

**Comment by psaghelyi** [comment]
Tested this branch (`feat/mcp-hooks-export` @ 548abd62) against our 14,902-drawer palace — posting here since @web3guru888 asked me on #478 to verify. **Partial fix**: the MCP server side lands cleanly, but the CLI still has the old bug.

## ✓ MCP server — correctly fixed

`mcp_server.py` adds `_fetch_all_metadata(col, where=None)` paginating in 1000-item batches, with a 5s TTL cache for repeated calls. All four tools — `tool_status`, `tool_list_wings`, `tool_list_rooms`, `tool_get_taxonomy` — now call it. Pattern matches `palace_graph.py:49-51`. Verified by reading the branch source rather than reconfiguring our running MCP server (didn't want to touch the live plugin install).

## ✗ CLI `mempalace status` — still broken

`miner.py:850-870` (`def status(palace_path)`) wasn't touched by this PR. Line 861 still reads:

```python
r = col.get(limit=10000, include=["metadatas"])
...
print(f"  MemPalace Status — {len(metas)} drawers")
```

Running the PR branch directly on our palace:

```bash
$ uvx --from 'git+https://github.com/jphein/mempalace@feat/mcp-hooks-export' mempalace status | head -3
=======================================================
  MemPalace Status — 10000 drawers
=======================================================
```

Same 11-of-17 wing truncation we saw on 3.1.0 — `askalot-analysis`, `docs`, `e2e-tests`, `infrastructure`, `scripts`, `shared-frontend` are all silently dropped. `col.count()` still reports 14,902.

## Suggested trivial follow-up

The CLI `status()` function is a tight ~20 lines and can call straight into the new helper (or duplicate the pattern if you'd rather not import from `mcp_server`):

```python
def status(palace_path: str):
    try:
        client = chromadb.PersistentClient(path=palace_path)
        col = client.get_collection("mempalace_drawers")
    except Exception:
        print(f"\n  No palace found at {palace_path}")
        return

    total = col.count()
    metas = []
    offset = 0
    while offset < total:
        batch = col.get(limit=1000, offset=offset, include=["metadatas"])
        metas.extend(batch["metadatas"])
        if not batch["metadatas"]:
            break
        offset += len(batch["metadatas"])

    wing_rooms = defaultdict(lambda: defaultdict(int))
    for m in metas:
        wing_rooms[m.get("wing", "?")][m.get("room", "?")] += 1

    print(f"\n{'=' * 55}")
    print(f"  MemPalace Status — {total} drawers")
    # ... rest unchanged
```

Happy to send a PR against your branch if that's easier than bundling it into this one. Otherwise #478 stays half-open since the user-visible CLI still reports the wrong count.

Everything else in the MCP server diff (`_fetch_all_metadata`, the cache, the search limit cap, the new read tools) looks good from a code-review standpoint — I didn't test those end-to-end.

**Comment by web3guru888** [comment]
Thanks @psaghelyi — really valuable field test on a real 14.9K palace.

Good catch on `status()` in miner.py. I reviewed the MCP server path (`_fetch_all_metadata()`, the tool registry, WAL guard) but didn't walk through the CLI surface in miner.py:850–870. That's on me — same underlying defect, different call site, should have been part of the same sweep.

Your patch is correct and clean: replace `col.get(limit=10000)` with the paginated loop, accumulate, then report `len(all_ids)`. It mirrors exactly how `_fetch_all_metadata()` was fixed on the server side, which is the right approach.

I'd suggest @jphein roll this in before merge. Shipping the PR with MCP server fixed but `status()` still capped at 10K would leave half of #478 open — and `status()` is the first thing ops people reach for when something looks wrong. If you're up for it @psaghelyi, a PR against `feat/mcp-hooks-export` is probably the fastest path.

**Comment by psaghelyi** [comment]
@jphein — rolled the `miner.status()` fix into a small PR against your feature branch: https://github.com/jphein/mempalace/pull/1

One hunk, mirrors the pagination pattern from your new `_fetch_all_metadata()` helper on the server side. Verified against our 14,902-drawer / 17-wing palace — CLI now reports the correct total and all wings. If you merge it into `feat/mcp-hooks-export` before this PR lands upstream, #478 closes cleanly on both the MCP and CLI surfaces in one go.

**Comment by jphein** [comment]
@psaghelyi — great catch and thanks for the thorough field test against your 14.9K palace. Merged your PR (#1) into `feat/mcp-hooks-export` — the CLI `status()` now paginates identically to the MCP server's `_fetch_all_metadata()`. Both surfaces of #478 are now fixed in this branch.

Also merged latest `main` into both PR branches (`fix/perf-quality` and `feat/mcp-hooks-export`) to pick up the `_MAX_RESULTS` constant extraction and test count update.

**Comment by web3guru888** [comment]
@jphein — the cross-branch coordination with psaghelyi is exactly the right approach. The `status()` CLI and MCP `_fetch_all_metadata()` being pagination-identical is important — having two surfaces of #478 that behave differently would be a persistent source of confusion.

Merging `main` into both branches to pick up the `_MAX_RESULTS` constant extraction is also good hygiene — keeps the constant single-sourced before the final merge. Looking forward to the combined PR landing.

**Comment by jphein** [comment]
Superseded by consolidated PR — all changes plus review feedback incorporated into a single clean PR from main.
